File: assets/ws/generate-ts-client/demo/lib/WSClientBase.ts
import { SimpleEventDispatcher } from 'strongly-typed-events';

import { RestClient } from './rest/RestClient';
import { HubClient } from './hub/HubClient';
import { IWSBase } from './base/IWSBase';
import { WSBase } from './base/WSBase';
import { IWSError } from './base/IWSError';

export class WSClientBase {
    //#region [ properties ]
    public readonly rest: RestClient;
    public readonly hub: HubClient;
    public readonly ws: IWSBase;
    //#endregion

    //#region  [ event ]
    public onWSError = new SimpleEventDispatcher<IWSError>();
    //#endregion

    //#region [ constructor ]
    constructor() {
        this.ws = new WSBase();
        this.hub = new HubClient(this.ws);
        this.rest = new RestClient(this.ws);
        this.ws.onNewSocketInstance.sub(() => {
            this.rest.initialize();
            this.hub.initialize();
        });
        this.ws.onConnectionChange.sub(value => {
            if (value) {
                this.hub.connect();
            }
        });

        const self = this;
        this.rest.onWSError.sub((x) => self.onWSError.dispatch(x));
        this.hub.onWSError.sub(x => self.onWSError.dispatch(x));
    }
    //#endregion
}

File: assets/ws/generate-cs-client/demo/src/template.api.wsclient/src/lib/ws/WSClientBase.cs
using nex.types;
using System;

namespace nex.ws
{
    public abstract class WSClientBase
    {
        #region [ properties ]
        public RestClient Rest { get; }
        public HubClient Hub { get; }
        public IWSBase    Ws { get; }
        #endregion

        #region [ events ]
        public event EventHandler<EventArgs<WSError>> EventWSError;
        #endregion

        #region [ constructor  ]
        public WSClientBase()
        {
            Ws = new WSBase();
            Hub = new HubClient(Ws);
            Rest = new RestClient(Ws);

            Ws.EventNewSocketInstance += (s, e) =>
            {
                Hub.Initialize();
                Rest.Initialize();
            };
            Ws.EventConnectionChange += (s, e) =>
            {
                if (e.Value)
                {
                    Hub.connect();
                }
            };

            Rest.EventWSError += (s, e) => EventWSError?.Invoke(this, e);
            Hub.EventWSError += (s, e) => EventWSError?.Invoke(this, e);
        }
        #endregion
    }
}
